src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,536 @@
+                {
+                    *table_term *= &theta;
+                    *table_term += fixed_value;
+                }
+            });
+        }
+
+        // Add blinding \beta and \gamma
+        parallelize(&mut lookup_product, |product, start| {
+            for ((product, input_term), table_term) in product
+                .iter_mut()
+                .zip(input_term[start..].iter())
+                .zip(table_term[start..].iter())
+            {
+                *product *= &(*input_term + &beta);
+                *product *= &(*table_term + &gamma);
+            }
+        });
+
+        // The product vector is a vector of products of fractions of the form
+        //
+        // (a_1(\omega^i) + \theta a_2(\omega^i) + ... + \beta)(s_1(\omega^i) + \theta s_2(\omega^i) + ... + \gamma)/
+        // (a'(\omega^i) + \beta) (s'(\omega^i) + \gamma)
+        //
+        // where a_j(\omega^i) is the jth input wire in this lookup,
+        // a'j(\omega^i) is the permuted input wire,
+        // s_j(\omega^i) is the jth table wire in this lookup,
+        // s'(\omega^i) is the permuted table wire,
+        // and i is the ith row of the wire.
+
+        // Compute the evaluations of the lookup product polynomial
+        // over our domain, starting with z[0] = 1
+        let mut z = vec![C::Scalar::one()];
+        for row in 1..(params.n as usize) {
+            let mut tmp = z[row - 1];
+            tmp *= &lookup_product[row];
+            z.push(tmp);
+        }
+        let z = pk.vk.domain.lagrange_from_vec(z);
+
+        // Check lagrange form of product is correctly constructed over the domain
+        // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
+        // - z'(\omega^{-1} X) (a_1(X) + \theta a_2(X) + ... + \beta) (s_1(X) + \theta s_2(X) + ... + \gamma)
+        let n = params.n as usize;
+
+        for i in 0..n {
+            let prev_idx = (n + i - 1) % n;
+
+            let mut left = z.get_values().clone()[i];
+            let permuted_input_value = &permuted.permuted_input_value.get_values()[i];
+
+            let permuted_table_value = &permuted.permuted_table_value.get_values()[i];
+
+            left *= &(beta + permuted_input_value);
+            left *= &(gamma + permuted_table_value);
+
+            let mut right = z.get_values().clone()[prev_idx];
+            let mut input_term = unpermuted_input_values
+                .iter()
+                .fold(C::Scalar::zero(), |acc, input| {
+                    acc * &theta + &input.get_values()[i]
+                });
+
+            let mut table_term = unpermuted_table_values
+                .iter()
+                .fold(C::Scalar::zero(), |acc, table| {
+                    acc * &theta + &table.get_values()[i]
+                });
+
+            input_term += &beta;
+            table_term += &gamma;
+            right *= &(input_term * &table_term);
+
+            assert_eq!(left, right);
+        }
```

Comment:
This test only checks intermediate representations within the `construct_product()` method, so it may be more useful for debugging than as a unit test. I've commented it out with a remark to that effect.